File: libs/etterlatte-behandling-model/src/main/kotlin/no/nav/etterlatte/libs/common/generellbehandling/GenerellBehandling.kt
package no.nav.etterlatte.libs.common.generellbehandling

import no.nav.etterlatte.libs.common.feilhaandtering.checkInternFeil
import no.nav.etterlatte.libs.common.sak.SakId
import no.nav.etterlatte.libs.common.tidspunkt.Tidspunkt
import java.util.UUID

data class Behandler(
    val saksbehandler: String,
    val tidspunkt: Tidspunkt,
)

data class Attestant(
    val attestant: String,
    val tidspunkt: Tidspunkt,
)

data class GenerellBehandling(
    val id: UUID,
    val sakId: SakId,
    val opprettet: Tidspunkt,
    val type: GenerellBehandlingType,
    val innhold: Innhold?,
    val tilknyttetBehandling: UUID? = null,
    val status: Status,
    val behandler: Behandler? = null,
    val attestant: Attestant? = null,
    val returnertKommenar: String? = null,
) {
    enum class Status {
        OPPRETTET,
        FATTET,
        RETURNERT,
        ATTESTERT,
        AVBRUTT,
    }

    fun kanEndres() =
        when (this.status) {
            Status.OPPRETTET, Status.RETURNERT -> true
            Status.AVBRUTT, Status.ATTESTERT, Status.FATTET -> false
        }

    fun erFattet() = this.status === Status.FATTET

    init {
        if (innhold !== null) {
            when (type) {
                GenerellBehandlingType.ANNEN ->
                    checkInternFeil(innhold is Innhold.Annen) {
                        "Type $type matcher " +
                            "ikke innhold navn: ${innhold.javaClass.simpleName}"
                    }
                GenerellBehandlingType.KRAVPAKKE_UTLAND ->
                    checkInternFeil(innhold is Innhold.KravpakkeUtland) {
                        "Type $type matcher ikke innhold navn: ${innhold.javaClass.simpleName}"
                    }
            }
        }
    }

    companion object {
        fun opprettFraType(
            type: GenerellBehandlingType,
            sakId: SakId,
        ) = GenerellBehandling(UUID.randomUUID(), sakId, Tidspunkt.now(), type, null, null, Status.OPPRETTET)

        fun opprettUtland(
            sakId: SakId,
            behandlingreferanse: UUID?,
        ) = GenerellBehandling(
            UUID.randomUUID(),
            sakId,
            Tidspunkt.now(),
            GenerellBehandlingType.KRAVPAKKE_UTLAND,
            null,
            tilknyttetBehandling = behandlingreferanse,
            status = Status.OPPRETTET,
        )
    }

    enum class GenerellBehandlingType {
        ANNEN,
        KRAVPAKKE_UTLAND,
    }
}

enum class GenerellBehandlingHendelseType {
    OPPRETTET,
    FATTET,
    ATTESTERT,
    UNDERKJENT,
}

File: apps/etterlatte-saksbehandling-ui/client/src/shared/types/Generellbehandling.ts
interface Behandler {
  saksbehandler: string
  tidspunkt: string
}

interface Attestant {
  attestant: string
  tidspunkt: string
}

export interface Generellbehandling {
  id: string
  sakId: number
  innhold: Innhold | null
  type: GenerellBehandlingType
  opprettet: string
  status: Status
  tilknyttetBehandling: string
  behandler?: Behandler
  attestant?: Attestant
  kommentar?: string
}

export function generellbehandlingErRedigerbar(status: Status): boolean {
  switch (status) {
    case Status.RETURNERT:
    case Status.OPPRETTET:
      return true
    case Status.FATTET:
    case Status.ATTESTERT:
    case Status.AVBRUTT:
      return false
  }
}

export enum Status {
  OPPRETTET = 'OPPRETTET',
  FATTET = 'FATTET',
  RETURNERT = 'RETURNERT',
  ATTESTERT = 'ATTESTERT',
  AVBRUTT = 'AVBRUTT',
}

export type GenerellBehandlingType = GenerellBehandlingKravpakkeUtlandType | GenerellBehandlingAnnenType

type GenerellBehandlingKravpakkeUtlandType = 'KRAVPAKKE_UTLAND'
type GenerellBehandlingAnnenType = 'ANNEN'

export type Innhold = KravpakkeUtland | Annen

export interface KravpakkeUtland {
  type: GenerellBehandlingKravpakkeUtlandType
  landIsoKode?: string[]
  dokumenter?: DokumentSendtMedDato[]
  rinanummer?: string
  begrunnelse?: string
}

export interface DokumentSendtMedDato {
  dokumenttype: string
  sendt: boolean
  dato?: string
}

export interface Annen {
  type: GenerellBehandlingAnnenType
  innhold: string
}
